src/lib/tests/test_kernels.cpp

```diff
@@ -0,0 +1,191 @@
+#include "gtest/gtest.h"
+#include <cassert>
+#include <iostream>
+#include <random>
+
+#include "arrops.h"  //  fancy approaches
+#include "vanilla.h" // a naive approach
+
+// constants
+#define BOXSIZE 10
+#define NCOORDS 3000
+
+int32_t ulpsDistance(const float a, const float b) {
+  // Save work if the floats are equal.
+  // Also handles +0 == -0
+  if (a == b)
+    return 0;
+
+  const auto max = std::numeric_limits<int32_t>::max();
+
+  // Max distance for NaN
+  if (isnan(a) || isnan(b))
+    return max;
+
+  // If one's infinite and they're not equal, max distance.
+  if (isinf(a) || isinf(b))
+    return max;
+
+  int32_t ia, ib;
+  memcpy(&ia, &a, sizeof(float));
+  memcpy(&ib, &b, sizeof(float));
+
+  // Don't compare differently-signed floats.
+  if ((ia < 0) != (ib < 0))
+    return max;
+
+  // Return the absolute value of the distance in ULPs.
+  int32_t distance = ia - ib;
+  if (distance < 0)
+    distance = -distance;
+  return distance;
+}
+
+int64_t ulpsDistance(const double a, const double b) {
+  // Save work if the floats are equal.
+  // Also handles +0 == -0
+  if (a == b)
+    return 0;
+
+  const auto max = std::numeric_limits<int64_t>::max();
+
+  // Max distance for NaN
+  if (isnan(a) || isnan(b))
+    return max;
+
+  // If one's infinite and they're not equal, max distance.
+  if (isinf(a) || isinf(b))
+    return max;
+
+  int64_t ia, ib;
+  memcpy(&ia, &a, sizeof(double));
+  memcpy(&ib, &b, sizeof(double));
+
+  // Don't compare differently-signed floats.
+  if ((ia < 0) != (ib < 0))
+    return max;
+
+  // Return the absolute value of the distance in ULPs.
+  int64_t distance = ia - ib;
+  if (distance < 0)
+    distance = -distance;
+  return distance;
+}
+
+template <typename T> inline void EXPECT_EQ_T(T result, T ref) {}
+
+template <> inline void EXPECT_EQ_T<float>(float result, float ref) {
+  EXPECT_FLOAT_EQ(result, ref);
+}
+
+template <> inline void EXPECT_EQ_T<double>(double result, double ref) {
+  EXPECT_DOUBLE_EQ(result, ref);
+}
+
+// creates nrandom floating points between 0 and limit
+template <typename T>
+void RandomFloatingPoint(T *target, const int nrandom, const int neglimit,
+                         const int poslimit) {
+  std::random_device
+      rd; // Will be used to obtain a seed for the random number engine
+  std::mt19937 gen(rd()); // Standard mersenne_twister_engine seeded with rd()
+  std::uniform_real_distribution<T> distribution(neglimit, poslimit);
+  for (size_t i = 0; i < nrandom; i++) {
+    target[i] = distribution(gen);
+  }
+}
+
+// coordinates class that is templated
+template <typename T> class Coordinates : public ::testing::Test {
+protected:
+  // members
+  int ncoords;
+  int nresults;
+  T *coords0 = nullptr;
+  T *coords1 = nullptr;
+  T *ref = nullptr;
+  T *results = nullptr;
+  T box[3];
+
+  // coordinates range from 0 - delta to BOXSIZE + delta
+  void InitCoords(const int n, const double boxsize, const double delta) {
```

Comment:
Consider taking `n_results` instead of `n_coords` as a parameter. This will mean that every value is valid and you won’t have to do a check.